api/odl.py

```diff
@@ -890,3 +956,509 @@ def _url_for(self, *args, **kwargs):
+            if response.status_code == 400:
+                # 400 is returned when the loan has been deleted. Leave this loan out of the result.
+                continue
+            feed = feedparser.parse(unicode(response.content))
+            entries = feed.get("entries")
+            if len(entries) < 1:
+                raise CirculationException()
+            entry = entries[0]
+            availability = entry.get("opds_availability", {})
+            if availability.get("status") != "available":
+                # This loan might be expired.
+                continue
+            start = datetime.datetime.strptime(availability.get("since"), self.TIME_FORMAT)
+            end = datetime.datetime.strptime(availability.get("until"), self.TIME_FORMAT)
+
+            activity.append(
+                LoanInfo(
+                    loan.license_pool.collection,
+                    loan.license_pool.data_source.name,
+                    loan.license_pool.identifier.type,
+                    loan.license_pool.identifier.identifier,
+                    start,
+                    end,
+                    external_identifier=loan.external_identifier,
+                )
+            )
+        for hold in holds:
+            info_url = hold.external_identifier
+            response = self._get(info_url, allowed_response_codes=["2xx", "3xx", "400"])
+            if response.status_code == 400:
+                # 400 is returned when the hold has been deleted. Leave this hold out of the result.
+                continue
+            feed = feedparser.parse(unicode(response.content))
+            entries = feed.get("entries")
+            if len(entries) < 1:
+                raise CirculationException()
+            entry = entries[0]
+            availability = entry.get("opds_availability", {})
+            if availability.get("status") not in ["ready", "reserved"]:
+                # This hold might be expired.
+                continue
+            start = datetime.datetime.strptime(availability.get("since"), self.TIME_FORMAT)
+            end = datetime.datetime.strptime(availability.get("until"), self.TIME_FORMAT)
+            position = entry.get("opds_holds", {}).get("position")
+
+            activity.append(
+                HoldInfo(
+                    hold.license_pool.collection,
+                    hold.license_pool.data_source.name,
+                    hold.license_pool.identifier.type,
+                    hold.license_pool.identifier.identifier,
+                    start,
+                    end,
+                    hold_position=position,
+                    external_identifier=hold.external_identifier,
+                )
+            )
+        return activity
+
+
+class SharedODLImporter(OPDSImporter):
+    NAME = SharedODLAPI.NAME
+
+    @classmethod
+    def _detail_for_elementtree_entry(cls, parser, entry_tag, feed_url=None):
```

Comment:
This is the sort of thing I was thinking when I wrote above about making `OPDSImporter` handle other uses of OPDS. I think you could reuse this code in the `SharedODLAPI` instead of parsing out links and delivery mechanisms.

Reply:
I can't reuse much since the logic is different for borrow links and generic acquisition links, but I can move more of the code to this class.

Reply:
This is good for now but I think we could eventually move to a reusable OPDS parser that turns the XML (or JSON) into an abstract data structure.